fix: Allow `detect_missing_libraries` only while building

MBerguer opened the pull request:
# What :computer: 
The purpose of this PR is to remove the possibility of using the `detect_missing_libraries` from the config toml file because it does not make sense to have it there. This flag works by showing a list of missing libraries. It does not output any bytecode, but there are some other scenarios where this could happen, so the best way to fix it is to remove this from all the commands (script, test, create), which is the same as disabling it by default in all places and enabling it only for the build. 


# Why :hand:
We were experiencing issues with this flag (#710) because it was enabled for all the available subcommands.
And the error we were getting down the line was nearly understandable. Now, the user will still get the problem eventually, but if he looks above in the stack trace, there will be a message explaining why this happened, so he can change the config file and continue.


# Evidence :camera:
![image](https://github.com/user-attachments/assets/fff8b670-eaee-428f-ae91-08240ab02fd1)

# Notes :memo:
- Foundry book has to be updated to reflect this change.

MBerguer committed:
Detect_missing_libraries on build

The purpouse of this PR is to remove the possiblitityto use the detect_missing_libraries from the config toml file, because it does not makes sense to havie it there.
The way this flags works, is to show a list of libraries that are missing, and it does not output any bytecode, but there some other scenarios where this could happen, so the best way to fix it is to remove this from all the commands (script, test, create) which is the same as disable it by default in all places and enabling it only for the build.
finally, if there is someone still using this flag in the config, we show a warning mentioning that this is not the way to go.

fix

elfedy commented on crates/cli/src/opts/build/zksync.rs:
```diff
@@ -169,6 +173,12 @@ impl ZkSyncArgs {
         );
 
+        if zksync.detect_missing_libraries {
+            // warn saying that The `detect_missing_libraries` option should not be used in toml
```
Comment is repeating the warning test so it might make sense to remove it as the warning itself is self explanatory .

elfedy commented on crates/cli/src/opts/build/zksync.rs:
```diff
@@ -85,7 +85,11 @@ pub struct ZkSyncArgs {
+    #[clap(
+        long = "zk-detect-missing-libraries-deprecated",
+        id = "zk-detect-missing-libraries-deprecated",
+        group = "depracated"
```
```suggestion
        group = "deprecated"
```

elfedy commented on crates/forge/bin/cmd/build.rs:
```diff
@@ -66,6 +66,12 @@ pub struct BuildArgs {
 
+    /// Detect missing libraries, instead of erroring
+    ///
+    /// Currently unused
```
I think the `Currently unused` legend can be removed (it is used in fact)

elfedy commented on crates/cli/src/opts/build/zksync.rs:
```diff
@@ -85,7 +85,11 @@ pub struct ZkSyncArgs {
     /// Currently unused
-    #[clap(long = "zk-detect-missing-libraries")]
+    #[clap(
+        long = "zk-detect-missing-libraries-deprecated",
```
Maybe we should just remove from here instead of deprecating and then use the option in build args to create the right zk project.

MBerguer replied:
This is an intentional hack I had to do to show the warning when users currently have this flag in the TOML file.
If we remove all track of the flag for the rest of the commands (besides `build`), we can't look for it in the code.

If you have an alternative for this without creating a new flag argument, please let me know 🙏

elfedy replied:
Imo we should not warn about this because the usage in toml is rare and probably unintended. No working project will have this in config because it will yield no bytecode. The behaviour will no be that this silently does nothing and then if there are missing libraries it will fail compilation, which should make users aware of the issue (and then they can run the thing with the cli option)

MBerguer replied:
You are right. I wanted to be extra cautious here, but the probability of someone having this running on a CI or something similar right now is very low. Let's go for the straigtforward solution

elfedy commented on crates/forge/bin/cmd/build.rs:
```diff
@@ -66,6 +66,12 @@ pub struct BuildArgs {
+    ///
+    /// Currently unused
+    #[clap(long = "zk-detect-missing-libraries", group = "build")]
+    pub detect_missing_libraries: bool,
```
How is this being used here? I think it will still use the value from config and ignore this one, but what we probably want is use this value and not have it in config.
We need to add an argument to [config_create_project](https://github.com/matter-labs/foundry-zksync/blob/cb40c350aa60092f7fcea24c76e5e5721e3ad8a3/crates/config/src/zksync.rs#L176) to pass this flag down to the compiler [settings](https://github.com/matter-labs/foundry-zksync/blob/cb40c350aa60092f7fcea24c76e5e5721e3ad8a3/crates/config/src/zksync.rs#L140)

MBerguer replied:
this was being used for the warning. No needed anymore

elfedy commented on crates/forge/bin/cmd/build.rs:
```diff
@@ -66,6 +66,12 @@ pub struct BuildArgs {
+    ///
+    /// Currently unused
+    #[clap(long = "zk-detect-missing-libraries", group = "build")]
+    pub detect_missing_libraries: bool,
```
```suggestion
    pub zk_detect_missing_libraries: bool,
```
Since this is a foundry struct that we are adding a zksync specific value to, we should prefix with `zk`. Also maybe move to the bottom of the struct definition to separate this from the upstream fields.

MBerguer replied:
I agree about the naming, but if we change this here, we will affect existing projects currently using `detect_missing_libraries`... right?

elfedy replied:
For those using the command line no because the value is with the `zk` prefix already. Also, this is the name of the struct field, not the user facing option for setting the flag. If we wanted to keep the non-prefixed value (don't think we should tbh), we can add an alias via the clap macro.

elfedy reviewed (commented):
Imo if we are removing the option from everything but build, we should remove it from shared config and only keep the build flag

MBerguer committed:
tests working before ripping off the warning

MBerguer committed:
No warning and only on build

- Removing the warning for all the commands besides build
- Cleaning up the three tests

MBerguer committed:
naming

MBerguer committed:
renaming to zk_***

zk_detect_missing_libraries since we have this naming convention

MBerguer committed:
unused imports

MBerguer committed:
passing a parameter

Jrigada committed:
Merge branch 'main' into tincho/detect-missing-libraries

MBerguer committed:
tamo

MBerguer committed:
dependencies

MBerguer committed:
Renaming + fixing test

Rename it back so the compiler understand
fixing the test to assert for the missing library

MBerguer committed:
cleanup

MBerguer commented:
Closing this in favor of https://github.com/matter-labs/foundry-zksync/issues/821.
See #722 for more details.
